src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -517,17 +547,33 @@ impl Net {
                 METRICS.mmds.tx_frames.inc();
                 METRICS.mmds.tx_bytes.add(len as u64);
                 init_vnet_hdr(&mut self.rx_frame_buf);
+                self.rx_buffer
+                    .iovec
+                    .write_all_volatile_at(&self.rx_frame_buf[..vnet_hdr_len() + len], 0)?;
+                self.rx_buffer
+                    .mark_used((vnet_hdr_len() + len).try_into().unwrap(), &self.metrics);
                 return Ok(vnet_hdr_len() + len);
             }
         }
 
-        self.read_tap().map_err(NetError::IO)
+        let len = self.read_tap().map_err(NetError::IO)?;
+        self.rx_buffer
+            .mark_used(len.try_into().unwrap(), &self.metrics);
+        Ok(len)
     }
 
+    /// Read as many frames as possible.
     fn process_rx(&mut self) -> Result<(), DeviceError> {
-        // Read as many frames as possible.
+        if self.rx_buffer.iovec.len() == 0 {
+            self.parse_rx_descriptors();
+        }
```

Comment:
We can always call `self.parse_rx_descriptors();`. This is way there is a chance for us to read more packets (as we can parse additional chains). Otherwise we always rely on `RX` event handling to do preprocessing.

Reply:
I want to avoid parsing descriptors in the hot path, when possible.

In my previous attempt to implement `readv` parsing descriptors the moment we want to actually read from the TAP was leading to performance degradation. It is true that parsing is faster now but chances are that we will see _some_ degradation again.

Typically, the guest provides us with 13 descriptors. So imagine a scenario where we currently have 12 parsed descriptors in the queue. This means that we can receive up to 12 frames and pay the cost of parsing only for the 13th one. If we always parse, we will be paying the cost of parsing for all 13 frames.

Reply:
This parsing will only happens once per TAP event. So in your example we will parse 13th descriptor here, but then we go into actual loop that reads from the TAP, so we will potentially read all 13 packets. Otherwise we will only be able to read 12 packets (from your example only 12 desc are parsed) because buffer will not be empty.

Reply:
> Otherwise we will only be able to read 12 packets (from your example only 12 desc are parsed) because buffer will not be empty.

I get that. What I'm saying is that I would like us to avoid reading if we don't need to. I will another (hopefully better) example. If we currently have 1 descriptor chain parsed and there is a single frame to read from the tap device, we will pay the cost of parsing 12 descriptors for reading a single packet.

Reply:
yeah, this is valid. But I think if TAP has small amount of packets to read, this means network is not under high load, so doing parsing here will not cause major delay. Last argument I have: we will need to always parse for mrg_ rxbuf support, otherwise there will be a lot of retransmissions.

Reply:
Ok, I launched an A/B test to compare the version with conditional parsing against one we are always parsing, and I didn't see a noticeable difference: https://buildkite.com/firecracker/bchalios-a-b-testing/builds/102#_. So I will implement this, but I will do it inside `read_from_mmds_or_tap` so that I will only parse extra descriptors when I need them (`rx_buffer.len() == 0`). This way we can read from the TAP immediately and only spend time to parse descriptors for the frames we actually need them for.